src/rtp_receiver.rs

```diff
@@ -4,34 +4,81 @@ use compositor_common::{
     scene::{InputId, Resolution},
     Frame,
 };
+use compositor_pipeline::{pipeline::PipelineInput, queue::Queue};
+use crossbeam_channel::{bounded, Sender};
 use log::{error, warn};
-use std::{fs::File, io::Write, path::PathBuf, sync::Arc, thread, time::Duration};
+use std::{
+    ffi::CString,
+    fs::File,
+    io::Write,
+    mem,
+    path::{Path, PathBuf},
+    ptr,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread,
+    time::Duration,
+};
 
 use ffmpeg_next::{
     codec::{self, Context, Id},
-    format,
+    ffi::{
+        avformat_alloc_context, avformat_close_input, avformat_find_stream_info,
+        avformat_open_input,
+    },
+    format::{self, context},
     frame::{self, Video},
     media::Type,
-    Dictionary,
+    util::interrupt,
+    Dictionary, Packet, Stream,
 };
 
-use crate::state::Pipeline;
-
-#[allow(dead_code)]
 pub struct RtpReceiver {
-    port: u16,
+    drop_sender: Sender<()>,
+    should_close: Arc<AtomicBool>,
 }
 
-impl RtpReceiver {
-    pub fn new(pipeline: Arc<Pipeline>, port: u16, input_id: InputId) -> Self {
-        let port_clone = port;
+pub struct Options {
+    pub port: u16,
+    pub input_id: InputId,
+}
+
+impl PipelineInput for RtpReceiver {
+    type Opts = Options;
+
+    fn new(queue: Arc<Queue>, opts: Self::Opts) -> Self {
+        let (drop_sender, drop_receiver) = bounded(0);
+        let should_close = Arc::new(AtomicBool::new(false));
+        let should_close_clone = should_close.clone();
         thread::spawn(move || {
-            RtpReceiver::start(pipeline, port_clone, input_id).unwrap();
+            RtpReceiver::start(queue, opts.port, opts.input_id, should_close_clone).unwrap();
+            drop_receiver.recv().unwrap();
         });
-        Self { port }
+        Self {
+            drop_sender,
+            should_close,
+        }
     }
+}
 
-    fn start(pipeline: Arc<Pipeline>, port: u16, input_id: InputId) -> Result<()> {
+impl Drop for RtpReceiver {
+    fn drop(&mut self) {
+        // - AtomicBool signals to RTP thread that it should abort
+        // - Channel signals to drop method that RTP thread finished
+        self.should_close.store(true, Ordering::Relaxed);
+        self.drop_sender.send(()).unwrap();
```

Comment:
a bit weird to use a channel for one way communication and an atomic for the other way, but it will work